statsmodels/graphics/functional.py

```diff
@@ -1,14 +1,441 @@
+        are computed. If `min_alpha` is not provided (single quantile value),
+        `max_pdf` is set to `1E6` in order not to constrain the problem on high
+        values.
+
+        An optimization is performed per component in order to find the min and
+        max curves. This is done by comparing the PDF value of a given curve
+        with the band PDF.
+
+        Parameters
+        ----------
+        band : array_like
+            alpha values ``(max_alpha, min_alpha)`` ex: ``[0.9, 0.5]``
+
+        Returns
+        -------
+        band_quantiles : list of 1-D array
+            ``(max_quantile, min_quantile)`` (2, n_features)
+
+        """
+        min_pdf = pvalues[alpha.index(band[0])]
+        try:
+            max_pdf = pvalues[alpha.index(band[1])]
+        except IndexError:
+            max_pdf = 1E6
+        band = [min_pdf, max_pdf]
+
+        pool = Pool()
+        data = zip(range(dim), itertools.repeat((band, pca,
+                                                 bounds, ks_gaussian)))
+        band_quantiles = pool.map(_min_max_band, data)
+        pool.terminate()
+        pool.close()
+
+        band_quantiles = list(zip(*band_quantiles))
+
+        return band_quantiles
+
+    extra_alpha = [i for i in alpha
+                   if 0.5 != i and 0.9 != i and threshold != i]
+    if extra_alpha != []:
+            extra_quantiles = [y for x in extra_alpha
+                               for y in _band_quantiles([x])]
+    else:
+        extra_quantiles = []
+
+    # Inverse transform from n-variate plot to dataset dataset's shape
+    median = _inverse_transform(pca, median)[0]
+    hdr_90 = _band_quantiles([0.9, 0.5])
+    hdr_50 = _band_quantiles([0.5])
+
+    hdr_res = HdrResults({
+                            "median": median,
+                            "hdr_50": hdr_50,
+                            "hdr_90": hdr_90,
+                            "extra_quantiles": extra_quantiles,
+                            "outliers": outliers
+                         })
+
+    # Plots
+    ax.plot(np.array([xdata] * n_samples).T, data.T,
+            c='c', alpha=.1, label='dataset')
+    ax.plot(xdata, median, c='k', label='Median')
+    ax.fill_between(xdata, *hdr_50,
+                    color='gray', alpha=.4,  label='50% HDR')
+    ax.fill_between(xdata, *hdr_90,
+                    color='gray', alpha=.3, label='90% HDR')
+
+    if len(extra_quantiles) != 0:
+        ax.plot(np.array([xdata] * len(extra_quantiles)).T,
+                np.array(extra_quantiles).T,
+                c='y', ls='-.', alpha=.4, label='Extra quantiles')
+
+    if len(outliers) != 0:
+        for ii, outlier in enumerate(outliers):
+            label = str(labels[ii]) if labels is not None else 'Outliers'
+            ax.plot(xdata, outlier,
+                    ls='--', alpha=0.7, label=label)
+
+    if labels is not None:
+        handles, labels = ax.get_legend_handles_labels()
+        by_label = dict(zip(labels, handles))
```

Comment:
AFAICS, the use of dict here makes the ordering of lables non-deterministic.
E.g. I don't like that labels in the legend use random mixup of outlier and stats, like HDR, labels

Reply:
I use this snippet to filter the labels and not have duplicate labels for the data part of the plot. If you have another way to deal with it I will be happy to change this.

Reply:
looks like using ordereddict fromkeys is now "state of the art"
https://stackoverflow.com/questions/480214/how-do-you-remove-duplicates-from-a-list-in-whilst-preserving-order/39835527#39835527
based on second answer here
https://stackoverflow.com/questions/480214/how-do-you-remove-duplicates-from-a-list-in-whilst-preserving-order

Reply:
Is this OK with you?
```
if labels is not None:
        handles, labels = ax.get_legend_handles_labels()
        by_label = OrderedDict(zip(labels, handles))
        ax.legend(by_label.values(), by_label.keys(), loc='best')
```
Order is preserved. In any case, for the dataset we have to select one handle from the list of handles.